modules/tile-converter/src/i3s-converter/helpers/node-index-document.ts

```diff
@@ -0,0 +1,276 @@
+import {join} from 'path';
+import {
+  BoundingVolumes,
+  LodSelection,
+  Node3DIndexDocument,
+  NodeInPage,
+  NodeReference
+} from '@loaders.gl/i3s';
+import transform from 'json-map-transform';
+import {v4 as uuidv4} from 'uuid';
+import {openJson, writeFile, writeFileForSlpk} from '../../lib/utils/file-utils';
+import I3SConverter from '../i3s-converter';
+import {NODE as nodeTemplate} from '../json-templates/node';
+import {I3SConvertedResources} from '../types';
+
+export class NodeIndexDocument {
+  public id: string;
+  public inPageId: number;
+  public data: Node3DIndexDocument | null = null;
+  public children: NodeIndexDocument[] = [];
+  private converter: I3SConverter;
+
+  constructor(id: number, converter: I3SConverter) {
+    this.inPageId = id;
+    this.id = id === 0 ? 'root' : id.toString();
+    this.converter = converter;
+  }
+
+  public async addData(data: Node3DIndexDocument): Promise<NodeIndexDocument> {
+    if (this.converter.options.instantNodesWriting) {
+      await this.write(data);
+    } else {
+      this.data = data;
+    }
+    return this;
```

Comment:
Is it worth chaining these? A promise that returns the class is not very useful... Just return void?